1단계 - 레거시 코드 리팩터링

developer-shkim opened the pull request:
# 작업 내용
- 기존 QnaService 에 있던 도메인 로직들을 Question 으로 옮기고, Question 에서 Answer 를 삭제하는 메소드를 호출하도록 수정했습니다.
- DeleteHistory 생성자를 private 으로 변경하고 Question, Answer 의 DeleteHistory 를 생성하는 팩토리 메서드를 추가했습니다.

리뷰 부탁드립니다! 감사합니다.

developer-shkim committed:
docs(lms): 요구사항 README 에 정리

developer-shkim committed:
refactor(lms): 1단계 - 레거시 코드 리팩터링

seondongpyo commented on src/main/java/nextstep/qna/domain/Answer.java:
```diff
@@ -76,4 +76,10 @@ public void toQuestion(Question question) {
         return "Answer [id=" + getId() + ", writer=" + writer + ", contents=" + contents + "]";
     }
+
+    public DeleteHistory delete() {
```
답변을 삭제하는 경우에도 로그인한 사용자와 답변 사용자가 동일한지 검증해야 하지 않을까요?
해당 메서드의 접근 제어자가 `public`으로 외부에 열려 있기 때문에, 누군가사 실수로 해당 메서드를 호출하게 된다면
검증 절차 없이 답변이 삭제가 되어버릴 수 있을 것 같습니다.

developer-shkim replied:
넵 답변 삭제 시 검증 로직 추가했습니다!

seondongpyo commented on src/main/java/nextstep/qna/domain/Question.java:
```diff
@@ -89,4 +85,29 @@ public List<Answer> getAnswers() {
+        List<DeleteHistory> deleteHistories = new ArrayList<>(
+                List.of(DeleteHistory.createQuestionHistory(id, writer, LocalDateTime.now())));
+
+        for (Answer answer : answers) {
+            deleteHistories.add(answer.delete());
+        }
```
`List<Answer> answers`를 일급 컬렉션으로 만들어서 처리해보면 어떨까요?

developer-shkim replied:
넵 일급 컬렉션으로 수정했습니다!

seondongpyo commented on src/test/java/nextstep/qna/domain/QuestionTest.java:
```diff
@@ -1,8 +1,53 @@
+
+    @Test
+    void delete_삭제_권한이_없는_사용자가_삭제하면_에러가_발생한다() {
+        CannotDeleteException exception = assertThrows(CannotDeleteException.class, () ->
+                question.delete(new NsUser()));
+        assertThat(exception.getMessage()).isEqualTo("질문을 삭제할 권한이 없습니다.");
```
Nit: `assertThatThrownBy()`를 활용해서도 테스트 가능한 점 참고해주시면 좋을 것 같습니다.
```suggestion
        assertThatThrownBy(() -> question.delete(new NsUser()))
            .isInstanceOf(CannotDeleteException.class)
            .hasMessage("질문을 삭제할 권한이 없습니다.");
```

developer-shkim replied:
감사합니다! 수정했습니다~

seondongpyo commented on src/test/java/nextstep/qna/domain/QuestionTest.java:
```diff
@@ -1,8 +1,53 @@
+
+
+    @Test
+    void delete_질문을_삭제하면_삭제상태가_true가_된다() throws CannotDeleteException {
```
질문에 속한 답변들의 삭제 상태들도 `true`로 변경이 되었는지도 검증해보면 어떨까요?

developer-shkim replied:
넵 추가했습니다!

한 가지 추가로 궁금한 점이 있는데요~
아래 의견에 대해 피드백 주시면 감사하겠습니다!

Question 클래스 테스트 코드에서는 Question 의 delete 메소드에서 수행하는 로직을 테스트하고, 해당 메소드에는 답변 삭제가 true 로 변경되는 것까지 관심을 두지 않는다고 생각했습니다. 그래서 답변 삭제 상태를 true 로 변경이 되었는지는 Answer 의 delete 메소드 테스트 코드에서 확인하도록 했었습니다.

관련 테스트를 추가한다면 answers.delete 함수 호출 여부를 추가하는 것이 어떨지 생각했습니다.

seondongpyo replied:
`Answers`의 `delete()` 메서드 호출 여부만으로는 Question의 `delete()`를 호출했을 때 반드시 질문에 속한 답변들의 삭제 상태가 `true`로 변경이 되었음을 테스트 코드에서 보장할 수는 없지 않을까 생각합니다.

developer-shkim replied:
넵 감사합니다!

seondongpyo reviewed (changes requested):
안녕하세요, 수현님. 수강신청 미션을 함께하게 된 표선동이라고 합니다.
1단계 리팩토링 미션 잘 진행해주셨네요 👍 
몇 가지 코멘트 드렸는데 확인 후 다시 리뷰 요청을 부탁 드리겠습니다 🔥

developer-shkim committed:
refactor(lms): 리뷰 대응

- 답변 삭제 시에도 사용자 검증 추가
- List<Answers> 를 일급 컬렉션으로 변경
- 테스트 코드 보완

seondongpyo commented on src/main/java/nextstep/qna/domain/Question.java:
```diff
@@ -65,28 +66,43 @@ public NsUser getWriter() {
     public void addAnswer(Answer answer) {
         answer.toQuestion(this);
-        answers.add(answer);
+
+        List<Answer> originAnswers = this.answers.getValue();
+        originAnswers.add(answer);
+
+        answers = new Answers(originAnswers);
```
`Answers`에서 값을 꺼내서 처리하는 대신에, 메시지를 보내서 처리해보면 어떨까요?
```suggestion
        answers.add(answer);
```

seondongpyo commented on src/test/java/nextstep/qna/domain/AnswersTest.java:
```diff
@@ -0,0 +1,39 @@
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+public class AnswersTest {
```
클래스의 이름은 `AnswersTest`인데, 실제로는 `Answers`가 아닌 `Answer`에 대한 단위 테스트만 작성된 것 같습니다.

seondongpyo commented on src/test/java/nextstep/qna/domain/QuestionTest.java:
```diff
@@ -1,8 +1,50 @@
+    void delete_질문을_삭제하면_답변_삭제상태가_true가_된다() throws CannotDeleteException {
+        question.delete(NsUserTest.JAVAJIGI);
+
+        question.getAnswers()
+                .getValue()
+                .forEach(answer -> assertThat(answer.isDeleted()).isTrue());
```
삭제하려는 질문에 답변이 하나도 없는 것 같은데, 실제로 답변을 추가하여 테스트를 진행해보면 어떨까요?

seondongpyo commented on src/test/java/nextstep/qna/domain/QuestionTest.java:
```diff
@@ -1,8 +1,50 @@
+    void delete_삭제하면_삭제_이력들을_반환한다() throws CannotDeleteException {
+        List<DeleteHistory> histories = question.delete(NsUserTest.JAVAJIGI);
+
+        assertThat(histories).isNotNull();
```
삭제 이력의 개수까지 검증해보면 어떨까요? (ex. 질문만 삭제하면 1개, 질문에 답변이 2개 있다면 총 3개)

seondongpyo reviewed (changes requested):
피드백 잘 반영해주셨네요 👍 
몇 가지 요구사항을 충족하지 않는 부분이 있어서 추가로 코멘트 드렸습니다.
확인 후 다시 리뷰 요청을 부탁 드리겠습니다 🔥

developer-shkim committed:
refactor(lms): 리뷰 대응

- answer 객체를 꺼내지 않고 answers 메서드 호출
- 누락된 AnswersTest 작성
- 답변 있는 질문에 대한 테스트 케이스 보완

developer-shkim commented:
리뷰 주신 아래 부분 수정햇습니다! 확인 부탁드립니다~

# 변경 사항
- answer 객체를 꺼내지 않고 answers 메서드 호출
- 누락된 AnswersTest 작성
- 답변 있는 질문에 대한 테스트 케이스 보완

seondongpyo reviewed (approved):
피드백 잘 반영해주셨네요 👍 
이번 리팩토링 단계는 여기서 merge 하겠습니다.
다음 단계에서 뵐게요 🔥
